src/qpx_mmc.cpp

```diff
@@ -0,0 +1,4244 @@
+  }
+  return 0;
+}
+
+int seek(drive_info *drive, int32_t lba, unsigned char flags) {
+  drive->cmd[0] = MMC_SEEK;
+  drive->cmd[2] = (lba >> 24) & 0xFF;
+  drive->cmd[3] = (lba >> 16) & 0xFF;
+  drive->cmd[4] = (lba >> 8) & 0xFF;
+  drive->cmd[5] = lba & 0xFF;
+
+  drive->cmd[9] = flags;
+  if ((drive->err = drive->cmd.transport(NONE, NULL, 0))) {
+    sperror("SEEK", drive->err);
+    return (drive->err);
+  }
+  return 0;
+}
+
+int read_cd(drive_info *drive, unsigned char *data, int32_t lba,
+            int sector_count, unsigned char flags, unsigned char FUA) {
+  //	int transfer_length = sector_count * 3072;
+
+  //	int sect_data = 2352;
+  int sect_data = 3072;
+
+  int transfer_length = sector_count * sect_data;
+  if (sector_count < 0)
+    return -1;
+
+  //	printf("lba: %d, cnt: %d\n", lba,sector_count);
+  drive->cmd[0] = MMC_READ_CD;
+  drive->cmd[1] = FUA ? 0x08 : 0x00;
+  drive->cmd[2] = (lba >> 24) & 0xFF;
+  drive->cmd[3] = (lba >> 16) & 0xFF;
+  drive->cmd[4] = (lba >> 8) & 0xFF;
+  drive->cmd[5] = lba & 0xFF;
+  drive->cmd[8] = sector_count;
+  drive->cmd[9] = flags;
+  if ((drive->err = drive->cmd.transport(READ, data, transfer_length))) {
+    sperror("READ_CD", drive->err);
+    return (drive->err);
+  }
+  return 0;
+}
+
+int read(drive_info *drive, unsigned char *data, int32_t lba, int sector_count,
+         unsigned char FUA) {
+  //	int transfer_length = sector_count * 3072;
+  int transfer_length = sector_count * 2048;
+  if (sector_count < 0)
+    return -1;
+  drive->cmd[0] = MMC_READ;
+  drive->cmd[1] = FUA ? 0x08 : 0x00;
+  drive->cmd[2] = (lba >> 24) & 0xFF;
+  drive->cmd[3] = (lba >> 16) & 0xFF;
+  drive->cmd[4] = (lba >> 8) & 0xFF;
+  drive->cmd[5] = lba & 0xFF;
+  drive->cmd[8] = sector_count;
+  if ((drive->err = drive->cmd.transport(READ, data, transfer_length))) {
+    sperror("READ", drive->err);
+    return (drive->err);
+  }
+  return 0;
+}
+
+int read_one_ecc_block(drive_info *drive, unsigned char *data, int32_t lba) {
+  drive->cmd[0] = MMC_READ;
+  drive->cmd[2] = (lba >> 24) & 0xFF;
+  drive->cmd[3] = (lba >> 16) & 0xFF;
+  drive->cmd[4] = (lba >> 8) & 0xFF;
+  drive->cmd[5] = lba & 0xFF;
+  drive->cmd[8] = 0x10;
+  if ((drive->err = drive->cmd.transport(READ, data, 0x8000))) {
+    sperror("READ_ONE_ECC_BLOCK", drive->err);
+    return (drive->err);
+  }
+  //	if ((drive->err=drive->cmd.transport(READ,(void*)((unsigned
+  //char*)drive->rd_buf+(0x0001<<14)),0x34))) 		{sperror
+  //("READ_ONE_ECC_BLOCK",drive->err); return (drive->err);}
+  return 0;
+}
+
+int get_drive_serial_number(drive_info *drive) {
+  //	char data[2048]; memset(data, 0, sizeof(data));
+  unsigned int data_length;
+  unsigned int length;
+  get_configuration(drive, FEATURE_LOGICAL_UNIT_SERIAL_NUMBER, &data_length,
+                    NULL);
+  if (drive->err)
+    return -1;
+  length = drive->rd_buf[11];
+  drive->rd_buf[12 + length] = 0;
+  if (data_length > 8)
+    strncpy(drive->serial, (char *)drive->rd_buf + 12, 16);
```

Comment:
Codacy found an issue: [Easily used incorrectly; doesn't always \0-terminate or check for invalid pointers [MS-banned] (CWE-120).  ](https://app.codacy.com/manual/jeremiahwilson777/qpxtool/pullRequest?prid=6191247)